specs/adagrams.spec.js

```diff
@@ -0,0 +1,143 @@
+
+    });
+  });
+
+  describe('scoreWord', () => {
+    const expectScores = (wordScores) => {
+      Object.entries(wordScores).forEach(([word, score]) => {
+        expect(Adagrams.scoreWord(word)).toBe(score);
+      });
+    };
+
+    it('returns an accurate numerical score according to the score chart', () => {
+      expectScores({
+        'A': 1,
+        'DOG': 5,
+        'WHIMSY': 17
+      });
+    });
+
+    it('returns a score regardless of the input case', () => {
+      expectScores({
+        'a': 1,
+        'dog': 5,
+        'wHiMsY': 17
+      });
+    });
+
+    it('returns a score of 0 if given an empty input', () => {
+      expectScores({
+        '': 0
+      });
+    });
+
+    it('adds an extra 8 points if word is 7 or more characters long', () => {
+      expectScores({
+        'XXXXXXX': 64,
+        'XXXXXXXX': 72,
+        'XXXXXXXXX': 80,
+        'XXXXXXXXXX': 88,
+      });
+    });
+  });
+
+  describe('highestScoreFrom', () => {
+    const scoredWord = (word) => ({ word, score: Adagrams.scoreWord(word) });
+
+    it('returns a hash that contains the word and score of best word in an array', () => {
+      const words = ['X', 'XX', 'XXX', 'XXXX'];
+      const correct = scoredWord('XXXX');
+
+      expect(Adagrams.highestScoreFrom(words)).toEqual(correct);
+    });
+
+    it('accurately finds best scoring word even if not sorted', () => {
+      const words = ['XXX', 'XXXX', 'X', 'XX'];
+      const correct = scoredWord('XXXX');
+
+      expect(Adagrams.highestScoreFrom(words)).toEqual(correct);
+    });
+
+    describe('in case of tied score', () => {
+      const expectTie = (words) => {
+        const scores = words.map(Adagrams.scoreWord);
+        const highScore = scores.reduce((h, s) => h < s ? s : h, 0);
+        const tiedWords = scores.filter((s) => s == highScore);
+
+        // Expect at least two tied words
+        expect(tiedWords.length).toBeGreaterThan(1);
+      };
+
+      it('selects the word with 10 letters', () => {
+        const words = ['AAAAAAAAAA', 'BBBBBB'];
+        const correct = scoredWord('AAAAAAAAAA');
+        expectTie(words);
+
+        expect(Adagrams.highestScoreFrom(words)).toEqual(correct);
+        expect(Adagrams.highestScoreFrom(words.reverse())).toEqual(correct);
+      });
+
+      it('selects the word with fewer letters when neither are 10 letters', () => {
+        const words = ['MMMM', 'WWW'];
+        const correct = scoredWord('WWW');
+        expectTie(words);
+
+        expect(Adagrams.highestScoreFrom(words)).toEqual(correct);
+        expect(Adagrams.highestScoreFrom(words.reverse())).toEqual(correct);
+      });
+
+      it('selects the first word when both have same length', () => {
+        const words = ['AAAAAAAAAA', 'EEEEEEEEEE'];
+        const first = scoredWord('AAAAAAAAAA');
+        const second = scoredWord('EEEEEEEEEE');
+        expectTie(words);
+
+        expect(Adagrams.highestScoreFrom(words)).toEqual(first);
```

Comment:
I'm looking at this code now... the `scoredWord` function could feel really distracting to me?

What would you feel like if the style of this code was more like:
```javascript
const words = ['AAAAAAAAAA', 'EEEEEEEEEE'];
const highestScoringWord = Adagrams.highestScoreFrom(words);
expect( highestScoringWord.word ).toEqual( words[0] );

const highestFromReverse = Adagrams.highestScoreFrom(words.reverse());
expect( highestFromReverse.word ).toEqual( words[1] );
```

?

The proposed structure is absolutely... more rudimentary, though, but I'm a little nervous about dropping them into this, I guess?

Reply:
I'm in general proposing removing the use of the `scoredWord` function, which I'm not sure we have time for anymore, though

Reply:
I don't see any reason we don't have time to take it out.